# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/angle/checkout/src/common/bitset_utils.h
# Commit: 3a91885d47e0
# Full Hash: 3a91885d47e0ef7fba9749562d52300eea67c651
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2019-11-11 21:52:52
# Regressor Bug: 1578576
# File Overlap Count: 4
# Description:
#   Bug 1578576 - Part 2: Update to ANGLE 3865 r=jgilbert
#   
#   Depends on D44561
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D44579
# ==============================================================================

diff -r 77cee77c0307 -r 3a91885d47e0 gfx/angle/checkout/src/common/bitset_utils.h
--- a/gfx/angle/checkout/src/common/bitset_utils.h	Fri Nov 08 16:41:42 2019 +0000
+++ b/gfx/angle/checkout/src/common/bitset_utils.h	Fri Nov 08 17:06:00 2019 +0000
@@ -21,6 +21,11 @@
 
 namespace angle
 {
+template <typename BitsT, typename ParamT>
+constexpr static BitsT Bit(ParamT x)
+{
+    return (static_cast<BitsT>(1) << static_cast<size_t>(x));
+}
 
 template <size_t N, typename BitsT, typename ParamT = std::size_t>
 class BitSetT final
@@ -128,14 +133,10 @@
     Iterator end() const { return Iterator(BitSetT()); }
 
   private:
-    constexpr static BitsT Bit(ParamT x)
-    {
-        return (static_cast<BitsT>(1) << static_cast<size_t>(x));
-    }
     // Produces a mask of ones up to the "x"th bit.
     constexpr static BitsT Mask(std::size_t x)
     {
-        return ((Bit(static_cast<ParamT>(x - 1)) - 1) << 1) + 1;
+        return ((Bit<BitsT>(static_cast<ParamT>(x - 1)) - 1) << 1) + 1;
     }
 
     BitsT mBits;
@@ -283,7 +284,7 @@
 template <size_t N, typename BitsT, typename ParamT>
 bool BitSetT<N, BitsT, ParamT>::test(ParamT pos) const
 {
-    return (mBits & Bit(pos)) != 0;
+    return (mBits & Bit<BitsT>(pos)) != 0;
 }
 
 template <size_t N, typename BitsT, typename ParamT>
@@ -330,7 +331,7 @@
 template <size_t N, typename BitsT, typename ParamT>
 BitSetT<N, BitsT, ParamT> &BitSetT<N, BitsT, ParamT>::operator^=(const BitSetT &other)
 {
-    mBits = (mBits ^ other.mBits) & Mask(N);
+    mBits = mBits ^ other.mBits;
     return *this;
 }
 
@@ -350,14 +351,14 @@
 template <size_t N, typename BitsT, typename ParamT>
 BitSetT<N, BitsT, ParamT> &BitSetT<N, BitsT, ParamT>::operator|=(BitsT value)
 {
-    mBits |= value;
+    mBits |= value & Mask(N);
     return *this;
 }
 
 template <size_t N, typename BitsT, typename ParamT>
 BitSetT<N, BitsT, ParamT> &BitSetT<N, BitsT, ParamT>::operator^=(BitsT value)
 {
-    mBits ^= value;
+    mBits ^= value & Mask(N);
     return *this;
 }
 
@@ -390,6 +391,7 @@
 template <size_t N, typename BitsT, typename ParamT>
 BitSetT<N, BitsT, ParamT> &BitSetT<N, BitsT, ParamT>::set()
 {
+    ASSERT(mBits == (mBits & Mask(N)));
     mBits = Mask(N);
     return *this;
 }
@@ -397,9 +399,10 @@
 template <size_t N, typename BitsT, typename ParamT>
 BitSetT<N, BitsT, ParamT> &BitSetT<N, BitsT, ParamT>::set(ParamT pos, bool value)
 {
+    ASSERT(mBits == (mBits & Mask(N)));
     if (value)
     {
-        mBits |= Bit(pos);
+        mBits |= Bit<BitsT>(pos) & Mask(N);
     }
     else
     {
@@ -411,6 +414,7 @@
 template <size_t N, typename BitsT, typename ParamT>
 BitSetT<N, BitsT, ParamT> &BitSetT<N, BitsT, ParamT>::reset()
 {
+    ASSERT(mBits == (mBits & Mask(N)));
     mBits = 0;
     return *this;
 }
@@ -418,13 +422,15 @@
 template <size_t N, typename BitsT, typename ParamT>
 BitSetT<N, BitsT, ParamT> &BitSetT<N, BitsT, ParamT>::reset(ParamT pos)
 {
-    mBits &= ~Bit(pos);
+    ASSERT(mBits == (mBits & Mask(N)));
+    mBits &= ~Bit<BitsT>(pos);
     return *this;
 }
 
 template <size_t N, typename BitsT, typename ParamT>
 BitSetT<N, BitsT, ParamT> &BitSetT<N, BitsT, ParamT>::flip()
 {
+    ASSERT(mBits == (mBits & Mask(N)));
     mBits ^= Mask(N);
     return *this;
 }
@@ -432,7 +438,8 @@
 template <size_t N, typename BitsT, typename ParamT>
 BitSetT<N, BitsT, ParamT> &BitSetT<N, BitsT, ParamT>::flip(ParamT pos)
 {
-    mBits ^= Bit(pos);
+    ASSERT(mBits == (mBits & Mask(N)));
+    mBits ^= Bit<BitsT>(pos) & Mask(N);
     return *this;
 }
 